lib/features/push_notification/domain/usecases/connect_web_socket_interactor.dart

```diff
@@ -0,0 +1,47 @@
+import 'dart:convert';
+
+import 'package:core/presentation/state/failure.dart';
+import 'package:core/presentation/state/success.dart';
+import 'package:core/utils/app_logger.dart';
+import 'package:dartz/dartz.dart';
+import 'package:jmap_dart_client/jmap/account_id.dart';
+import 'package:jmap_dart_client/jmap/core/session/session.dart';
+import 'package:jmap_dart_client/jmap/push/state_change.dart';
+import 'package:tmail_ui_user/features/push_notification/domain/repository/web_socket_repository.dart';
+import 'package:tmail_ui_user/features/push_notification/domain/state/web_socket_push_state.dart';
+
+class ConnectWebSocketInteractor {
+  final WebSocketRepository _webSocketRepository;
+
+  ConnectWebSocketInteractor(this._webSocketRepository);
+
+  Stream<Either<Failure, Success>> execute(
+    Session session,
+    AccountId accountId
+  ) async* {
+    try {
+      yield Right(InitializingWebSocketPushChannel());
+      yield* _webSocketRepository
+        .getWebSocketChannel(session, accountId)
+        .map(_toStateChange);
+    } catch (e) {
+      logError('ConnectWebSocketInteractor::execute: $e');
+      yield Left(WebSocketConnectionFailed(exception: e));
+    }
+  }
+
+  Either<Failure, Success> _toStateChange(dynamic data) {
+    if (data is String) {
+      data = jsonDecode(data);
+    }
+    StateChange? possibleStateChange;
+    try {
+      possibleStateChange = StateChange.fromJson(data);
+      return Right(WebSocketPushStateReceived(possibleStateChange));
+    } catch (e) {
+      logError('ConnectWebSocketInteractor::_toStateChange: '
+        'websocket message is not StateChange: $data');
+      return Right(WebSocketPushStateReceived(null));
```

Comment:
When pinging websocket server, a message will returned back to confirm the ping success. We should not return failure here as it will terminate the connection every time we ping

Reply:
So we still have to consider it a failure case and we have to check what the exception is to handle it next.

Reply:
How about we try to parse it to `Echo` object. If it pass, it's not a failure case?

Reply:
So please give it the correct state name.

Reply:
Done